src/circuit/gadget/ecc/chip/add.rs

```diff
@@ -0,0 +1,272 @@
+    let c = config.extras[4]; // Used for both C and gamma
+    let d = config.lambda.1; // Used for both D and delta
+    let lambda = config.lambda.0;
+
+    // Assign A, B, C, D, α, β, γ, δ
+    {
+        x_p.zip(x_q)
+            .zip(y_p)
+            .zip(y_q)
+            .map(|(((x_p, x_q), y_p), y_q)| -> Result<(), Error> {
+                if x_q == x_p {
+                    // x_q = x_p ⟹ A
+                    region.assign_advice(|| "set A", a, offset, || Ok(C::Base::one()))?;
+
+                    // Doubling case, λ = (y_q − y_p) / (x_q − x_p)
+                    if y_p != C::Base::zero() {
+                        let lambda_val = C::Base::from_u64(3)
+                            * x_p
+                            * x_p
+                            * (C::Base::from_u64(2) * y_p).invert().unwrap();
+                        region.assign_advice(|| "set lambda", lambda, offset, || Ok(lambda_val))?;
+                    }
+                } else {
+                    // α = 1 / (x_q - x_p)
+                    let alpha_val = (x_q - x_p).invert().unwrap();
+                    region.assign_advice(|| "set alpha", a, offset + 1, || Ok(alpha_val))?;
+
+                    // Non-doubling case, λ = (y_q - y_p) / (x_q - x_p)
+                    let lambda_val = (x_q - x_p).invert().unwrap() * (y_q - y_p);
+                    region.assign_advice(|| "set lambda", lambda, offset, || Ok(lambda_val))?;
+                }
+
+                if x_p == C::Base::zero() {
+                    // x_p = 0 ⟹ B
+                    region.assign_advice(|| "set B", b, offset, || Ok(C::Base::one()))?;
+                } else {
+                    // β = 1 / x_p
+                    let beta_val = x_p.invert().unwrap();
+                    region.assign_advice(|| "set beta", b, offset + 1, || Ok(beta_val))?;
+                }
+                if x_q == C::Base::zero() {
+                    // x_q = 0 ⟹ C
+                    region.assign_advice(|| "set C", c, offset, || Ok(C::Base::one()))?;
+                } else {
+                    // γ = 1 / x_q
+                    let gamma_val = x_q.invert().unwrap();
+                    region.assign_advice(|| "set gamma", c, offset + 1, || Ok(gamma_val))?;
+                }
+
+                if y_p == -y_q {
+                    // y_p = -y_p ⟹ D
+                    region.assign_advice(|| "set D", d, offset, || Ok(C::Base::one()))?;
+                } else {
+                    // δ = 1 / (y_q + y_p)
+                    let delta_val = (y_q + y_p).invert().unwrap();
+                    region.assign_advice(|| "set delta", d, offset + 1, || Ok(delta_val))?;
+                }
+                Ok(())
+            });
+    }
+
+    // Compute R = P + Q
+    let r = x_p
+        .zip(y_p)
+        .zip(x_q)
+        .zip(y_q)
+        .map(|(((x_p, y_p), x_q), y_q)| {
+            let p = C::from_xy(x_p, y_p).unwrap();
+            let q = C::from_xy(x_q, y_q).unwrap();
+            p + q
```

Comment:
If either p or q are points at infinity then this will fail to `unwrap()`. Check if they're points at infinity first (aka `(0, 0)` in the circuit) and then if they are do `C::zero()` or `C::identity()` (forget which one it is).

This is slightly confusing because the curve implementation rejects `(0, 0)` as a point at infinity (such a point does not exist on the affine plane) but inside the circuit we represent the point at infinity as `(0, 0)`.